Fix list's connect

Liklainy opened the pull request:
Predicate in the `Connect` method doesn't work as expected.
Internal `Filter` is not initialized properly when there were items in the list.
So removing items in some cases can cause an out of bounds exception in `Transform`.
There is a fast fix, which moves `Filter` outside.

Liklainy commented:
@RolandPheasant, can you, please, assist on this?
There is a test in the first commit which fails.
I think, the correct fix is to remove predicate from `Connect`, but it will break existing API.

RolandPheasant commented:
The original API having the filter provides an optimisation for large record sets. I know for sure some people depend on this. I wonder whether the issue can be fixed in the filter and transform methods.

Liklainy commented:
Will it be a better way to add new special filter without transform?
Which will not track item's IsMatch, and it will only check predicate when item is added.

RolandPheasant commented:
It probably would do. Accounting for every permutation is what makes DD so difficult to implement correctly.

Liklainy commented:
Added `FilterStatic`
Based on Filter, but I removed Refresh case and changed Replace.
Predicated is checked when item is added or replaced.

Is it enough?
Or there is a need to save `filtered` items by wrapping `ItemWithIndex` from the base list?
(to change by index and not by item)

RolandPheasant commented:
I am inclined to leave it as simple as possible. At least for now anyway.

I will sit on this for a day or two before merging as dealing with list collections, it is very easy to omit permutations.  That way I will have time to think.

Additionally I intend to have a round of optimisation for the list side of DD soon and I think filter is a prime candidate for me to look at with Greg's eyes.
